python/huggingfaceserver/huggingfaceserver/model.py

```diff
@@ -0,0 +1,244 @@
+# Copyright 2024 The KServe Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import uuid
+from threading import Thread
+
+from torch import Tensor
+
+from kserve.model import PredictorConfig
+from kserve.protocol.rest.v2_datamodels import GenerateRequest, GenerateResponse, TokenOutput
+from .async_generate_stream import AsyncGenerateStream
+from .task import ARCHITECTURES_2_TASK, MLTask
+from kserve.logging import logger
+import pathlib
+from typing import Dict, Union, Any, AsyncIterator, Optional
+
+from kserve.errors import InferenceError
+from kserve.storage import Storage
+
+from kserve.protocol.infer_type import InferRequest, InferResponse, InferInput
+from kserve.utils.utils import get_predict_response, get_predict_input
+from kserve import Model
+import torch
+
+try:
+    from vllm.sampling_params import SamplingParams
+    from vllm.vllm_async_engine import AsyncLLMEngine
+
+    _vllm = True
+except ImportError:
+    _vllm = False
+from transformers import AutoModelForCausalLM, AutoModelForSeq2SeqLM, AutoTokenizer, \
+    AutoConfig, \
+    AutoModelForSequenceClassification, AutoModelForTokenClassification, AutoModelForQuestionAnswering, \
+    AutoModelForMaskedLM, BatchEncoding, TensorType
+
+
+class HuggingfaceModel(Model):  # pylint:disable=c-extension-no-member
+    def __init__(self, model_name: str, kwargs,
+                 engine_args=None, predictor_config: Optional[PredictorConfig] = None):
+        super().__init__(model_name, predictor_config)
+        if kwargs is None:
+            kwargs = {}
+        self.kwargs = {}
+        tp_degree = kwargs.get('tensor_parallel_degree', -1)
+        self.device = torch.device(
+            "cuda" if torch.cuda.is_available() else "cpu"
+        )
+        if "device_map" in kwargs:
+            self.kwargs["device_map"] = kwargs['device_map']
+        elif tp_degree > 0:
+            self.kwargs["device_map"] = "auto"
+            world_size = torch.cuda.device_count()
+            assert world_size == tp_degree, f"TP degree ({tp_degree}) doesn't match available GPUs ({world_size})"
+        self.model_id = kwargs.get('model_id', None)
+        self.model_dir = kwargs.get('model_dir', None)
+        self.do_lower_case = kwargs.get('do_lower_case', True)
+        self.add_special_tokens = kwargs.get('add_special_tokens', True)
+        self.max_length = kwargs.get('max_length', None)
```

Comment:
Is it better to set this to a non `None` default value?

Reply:
Hmm seems like there are both max_length and max_new_tokens and max length is going to be deprecated 
https://huggingface.co/docs/transformers/main_classes/text_generation

https://discuss.huggingface.co/t/confused-about-max-length-and-max-new-tokens/30892

Reply:
yes, `max_new_tokens` is recommended now

Reply:
Wait this arg i used for tokenizer and the max_length  default is none 

https://huggingface.co/transformers/v2.11.0/main_classes/tokenizer.html#:~:text=max_length%20(%20int%20%2C%20optional%20%2C%20defaults,the%20model%20with%20max_length%20%3D%20tokenizer.

Reply:
Oh right, I confused it to the GenerationConfig param. This looks good!

Reply:
It makes sense that they deprecated this for generate config to avoid confusion.

Reply:
So we can add max_new_token when you work on vllm